proposed/2022/pack-referenced-projects.md

```diff
@@ -0,0 +1,71 @@
+# Pack referenced projects
+
+- Author Name [maxkoshevoi](https://github.com/maxkoshevoi)
+- Start Date (2022-03-28)
+- GitHub Issue [3891](https://github.com/NuGet/Home/issues/3891)
+- GitHub PR (-)
+
+## Summary
+
+Allow referenced projects' artifacts to be packed inside the NuGet.
+
+## Motivation
+
+When SDK-style project is being packed as a NuGet, all projects that it references are considered as NuGet themselves that this NuGet depends on.
+This is not always a desired behavior, sometimes user needs artifacts from referenced project(s) to be packed into the NuGet.
+Currently there's no easy way to override this behavior except for ditching `csproj`, and going back to `nuspec`. In many cases that involves piping lots of information from the `MSBuild` context into the `nuspec` evaluation via `key=value` parameters.  
+Chaining together dozens of arguments to achieve parity with metadata usually automagically introduced by the SDK becomes tedious, which eventually weakens the otherwise rich set of information attached to a modern NuGet package.
+
+## Explanation
+
+### Functional explanation
+
+Proposed solution is to add new `Pack` and `PackagePath` properties to `ProjectReference` (`<ProjectReference Include="..." Pack="True" PackagePath="tools" />`) and `PackageReference`. 
+If `Pack` is `False` or not present - consider the project (or package) as a NuGet's dependency, if it's `True` - pack its artifacts in.
+If `PackagePath` is not present - pack artifacts to the default location within the package. If it's present - pack them to the specified path.
+
+For `PackageReference`s it's useful in case referenced package is from private feed, but NuGet that's being packed will be public.
+
+### Technical explanation
+
+When NuGet is being created from `csproj` file:
+
+1) If `ProjectReference` doesn't have `Pack` property, or has it set to `False` - preserve behavior that have existed before introduction of this feature (don't include artifacts of referenced projects into the NuGet, and consider them as NuGet-dependencies).
+2) If `ProjectReference` has `Pack` property and it's set to `True` - include it's artifacts into the NuGet (to the default location or one specified in `PackagePath`).
+3) Do the same steps for `PackageReference`s
+
+## Drawbacks
+
+None
+
+## Rationale and alternatives
+
+- Use `IsPackable` property from referenced project to determine it it's supposed to be packed or be a NuGet dependency.
+
+It will be more automated, but won't fit some edge cases (also it will be a breaking change). And it isn't applicable to `PackageReferences`.
+
+- Introduce new project-wide `PackReferencedProjects` property that would switch pack behavior for all `ProjectReference`s inside a single project.
+
+This one is less flexible since it cannot switch on per-`ProjectReference` basis. And it isn't applicable to `PackageReferences`.
+
+- Another solution is to introduce a `IncludeReferencedProjects` switch for `nuget` CLI.
```

Comment:
Would this be for `dotnet` CLI instead?

Reply:
If a Project is being included into a joined package and never gets packed itself, it's dependencies would need to be something that has to be represented in the joined package.
Example:

Project 1 -> Project 2 -> Package A.

Say, Project 1 includes the reference project, and it is intended to be redistributed as a library, it's extremely likely that Project 1 needs Package A as a dependency.
This is a `reverse restore`. Instead of given the dependency, give me a resolved graph, it starts with a resolved graph and it tries to generate the correct top level dependencies. Note that we'd need to account for package include/excludes as well.

Reply:
> Would this be for `dotnet` CLI instead?

Correct. It'd be for dotnet CLI and the msbuild pack target (which the dotnet pack is just a wrapper for)